pkg/expression/builtin_compare.go

```diff
@@ -119,18 +119,19 @@ func (c *coalesceFunctionClass) getFunction(ctx sessionctx.Context, args []Expre
 		return nil, err
 	}
 
-	fieldTps := make([]*types.FieldType, 0, len(args))
+	flag := uint(0)
```

Comment:
Why need to handle this NotNullFlag here? I think `InferType4ControlFuncs` already handles this flag?

Reply:
> Why need to handle this NotNullFlag here? I think `InferType4ControlFuncs` already handles this flag?

After check, I think `InferType4ControlFuncs` does not handle this flag.

Reply:
`InferType4ControlFuncs` is a common type infer function that aims at aggregating some input types into one "common return type", it should be used by all the functions whose return type has this pattern, for example: `ifnull/case when/coalesce`, if you need add code to handle `NotNullFlag` in `coalesce`, you should double check the following question:
- Why you need to add special code to handle `NotNullFlag` in `coalesce` only? Will the problem you meet also happens in other functions that use `InferType4ControlFuncs` as its type infer function?